rand_distr/src/poisson.rs

```diff
@@ -137,15 +140,15 @@ where Standard: Distribution<N>
     }
 }
 
-impl<N: Float> Distribution<u64> for Poisson<N>
-where Standard: Distribution<N>
-{
-    #[inline]
-    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> u64 {
-        let result: N = self.sample(rng);
-        result.to_u64().unwrap()
-    }
-}
+// impl<N: Float + FloatConst> Distribution<u64> for Poisson<N>
+// where Standard: Distribution<N>
+// {
```

Comment:
Because the `Float` trait comes from an external crate, we can't use such impl anymore. :(

Reply:
What's the error here?

Reply:
```
error[E0119]: conflicting implementations of trait `rand::distributions::Distribution<u64>` for type `poisson::Poisson<u64>`:
   --> rand_distr/src/poisson.rs:143:1
    |
79  | / impl<N: Float + FloatConst> Distribution<N> for Poisson<N>
80  | | where Standard: Distribution<N>
81  | | {
82  | |     #[inline]
...   |
140 | |     }
141 | | }
    | |_- first implementation here
142 | 
143 | / impl<N: Float + FloatConst> Distribution<u64> for Poisson<N>
144 | | where Standard: Distribution<N>
145 | | {
146 | |     #[inline]
...   |
150 | |     }
151 | | }
    | |_^ conflicting implementation for `poisson::Poisson<u64>`
    |
    = note: upstream crates may add a new impl of trait `num_traits::float::Float` for type `u64` in future versions
    = note: upstream crates may add a new impl of trait `num_traits::float::FloatConst` for type `u64` in future versions
```
I think we can define a blanket casting impl in `rand` over float distributions, but I am not sure if such approach is a correct one.

Reply:
There is an alternative: directly implement for `f32`, `f64` and `u64` types.

The `u64` impl was added [here](https://github.com/rust-random/rand/pull/795) without a rationale. @vks what do you think (do we want to keep this impl)?

Reply:
Are you talking about the `f64` impl? A rationale was provided in the commit message:
> Internally, we generate `f64`, so it makes sense to let the user access that result directly, instead of forcing a conversion from `u64`.

The `u64` impl is the original one and existed before #795.

Reply:
In this case, since we are doing a breaking change anyway, I think we can remove the `Distribution<u64>` impl in favor of `Distribution<f64>`.

Reply:
In principle, I'm fine with removing the `u64` impl. However, it is currently nice to have, because casting `f64` to `u64` can introduce undefined behavior, which the `u64` impl avoided. In fact, this PR falls back to `f64 as u64` casts with potential undefined behavior, which the previous implementation avoided.

Reply:
For now I've removed the `u64` impl and associated tests. Having a potentially panicking code without any misuse in user code seems far from ideal, so I think it will be better to bubble correct casting responsibility to users.